### ISSUE TITLE
Autoencoder seems to not work at all

### ISSUE THREAD

**Issue by snurkabill** [fchollet/keras#1604]
Hi, I think I may have found issue when autoencoder is used. 

Consider simple code: 

```
(X_train, y_train), (X_test, y_test) = mnist.load_data()

X_train = X_train.reshape(60000, 784)
X_test = X_test.reshape(10000, 784)

Y_train = np_utils.to_categorical(y_train, 10)
Y_test = np_utils.to_categorical(y_test, 10)

X_train = X_train.astype('float32')
X_test = X_test.astype('float32')
X_train /= 255
X_test /= 255

batch_size = 128
nb_pretraining_epochs = 10
nb_epoch = 5

encoder = Dense(output_dim = 100, input_dim = 784)
decoder = Dense(output_dim = 784, input_dim = 100)
# encoder = containers.Sequential([Dense(200, input_shape = (784, ))])
# decoder = containers.Sequential([Dense(784, input_shape = (200, ))])

autoencoder = Sequential()
autoencoder.add(AutoEncoder(encoder = encoder,
                            decoder = decoder,
                            output_reconstruction = False))

autoencoder.compile(optimizer = SGD(lr = 0.001, momentum = 0.1), loss = 'mse')

autoencoder.fit(X = X_train,
                y = X_train,
                batch_size = batch_size,
                nb_epoch = nb_pretraining_epochs,
                verbose = 1)


```

results of the code above: 

Epoch 1/10
60000/60000 [==============================] - 1s - loss: 0.1580     
Epoch 2/10
60000/60000 [==============================] - 1s - loss: 0.1580     
Epoch 3/10
60000/60000 [==============================] - 0s - loss: 0.1580     
Epoch 4/10
60000/60000 [==============================] - 1s - loss: 0.1580     
Epoch 5/10
60000/60000 [==============================] - 1s - loss: 0.1580     
Epoch 6/10
60000/60000 [==============================] - 1s - loss: 0.1580     
Epoch 7/10
60000/60000 [==============================] - 0s - loss: 0.1580     
Epoch 8/10
60000/60000 [==============================] - 0s - loss: 0.1580     
Epoch 9/10
60000/60000 [==============================] - 1s - loss: 0.1580     
Epoch 10/10
60000/60000 [==============================] - 0s - loss: 0.1580 

when I print weights, there are no updates at all, so autoencoder did nothing with parameters. 

using theano backend, ubuntu 15.4.

**Comment by fchollet** [comment]
You're right. Let me push a fix.

I was never a fan of the Autoencoder API. It's currently a huge mess. Let's fix it once and for all...

**Comment by snurkabill** [comment]
Thanks, 

why are you not a fan of AE? Are you using RBMs?

**Comment by antoniosehk** [comment]
I run your code and it seems perfect... Here is my results:

Epoch 1/10
60000/60000 [==============================] - 4s - loss: 0.1501     
Epoch 2/10
60000/60000 [==============================] - 4s - loss: 0.1452      
Epoch 3/10
60000/60000 [==============================] - 4s - loss: 0.1413      
Epoch 4/10
60000/60000 [==============================] - 4s - loss: 0.1383      
Epoch 5/10
60000/60000 [==============================] - 4s - loss: 0.1358      
Epoch 6/10
60000/60000 [==============================] - 4s - loss: 0.1338      
Epoch 7/10
60000/60000 [==============================] - 4s - loss: 0.1321      
Epoch 8/10
60000/60000 [==============================] - 4s - loss: 0.1306     
Epoch 9/10
60000/60000 [==============================] - 4s - loss: 0.1293     
Epoch 10/10
60000/60000 [==============================] - 4s - loss: 0.1281

**Comment by fchollet** [comment]
Done and pushed. There is a new API, which now makes sense (which is a big improvement):

```python
    from keras.layers import containers

    # input shape: (nb_samples, 32)
    encoder = containers.Sequential([Dense(16, input_dim=32), Dense(8)])
    decoder = containers.Sequential([Dense(16, input_dim=8), Dense(32)])

    autoencoder = Sequential()
    autoencoder.add(AutoEncoder(encoder=encoder, decoder=decoder,
                                output_reconstruction=True))

    # training the autoencoder:
    autoencoder.compile(optimizer='sgd', loss='mse')
    autoencoder.fit(X_train, X_train, nb_epoch=10)

    # predicting compressed representations of inputs:
    autoencoder.output_reconstruction = False  # the autoencoder has to be recompiled after modifying this property
    autoencoder.compile(optimizer='sgd', loss='mse')
    representations = autoencoder.predict(X_test)

    # the model is still trainable, although it now expects compressed representations as targets:
    autoencoder.fit(X_test, representations, nb_epoch=1)  # in this case the loss will be 0, so it's useless

    # to keep training against the original inputs, just switch back output_reconstruction to True:
    autoencoder.output_reconstruction = True
    autoencoder.compile(optimizer='sgd', loss='mse')
    autoencoder.fit(X_train, X_train, nb_epoch=10)
    ```

**Comment by snurkabill** [comment]
@antoniosehk that is weird. Haven't you tried it after @fchollet pushed changes?

**Comment by antoniosehk** [comment]
No. I tried it before @fchollet  pushed the changes.

**Comment by fchollet** [comment]
You were probably on an outdated version. Anyway, use the new API, now.

**Comment by antoniosehk** [comment]
@fchollet, yes...I have not updated for a long time....
